ffsync/rds_locker.go

```diff
@@ -0,0 +1,182 @@
+package ffsync
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/featureform/fferr"
+	"github.com/featureform/helpers"
+	"github.com/google/uuid"
+	"github.com/jackc/pgx/v4/pgxpool"
+	_ "github.com/lib/pq"
+)
+
+type rdsKey struct {
+	id   string
+	key  string
+	done *chan error
+}
+
+func (k rdsKey) ID() string {
+	return k.id
+}
+
+func (k rdsKey) Key() string {
+	return k.key
+}
+
+func NewRDSLocker(config helpers.RDSConfig) (Locker, error) {
+	const tableName = "ff_locks"
+
+	db, err := helpers.NewRDSPoolConnection(config)
+	if err != nil {
+		return nil, err
+	}
+
+	connection, err := db.Acquire(context.Background())
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to acquire connection from the database pool: %w", err))
+	}
+
+	err = connection.Ping(context.Background())
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to ping the database: %w", err))
+	}
+
+	// Create a table to store the locks
+	tableCreationSQL := fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (lock_id VARCHAR(255), resource_key VARCHAR(255) NOT NULL, lock_timestamp TIMESTAMP NOT NULL DEFAULT CURRENT_TIMESTAMP, PRIMARY KEY (resource_key));", tableName)
+	_, err = db.Exec(context.Background(), tableCreationSQL)
+	if err != nil {
+		return nil, fferr.NewInternalError(fmt.Errorf("failed to create table %s: %w", tableName, err))
+	}
+
+	return &rdsLocker{
+		db:         db,
+		tableName:  tableName,
+		connection: connection,
+	}, nil
+}
+
+type rdsLocker struct {
+	db         *pgxpool.Pool
+	tableName  string
+	connection *pgxpool.Conn
+}
+
+func (l *rdsLocker) Lock(key string) (Key, error) {
+	if key == "" {
+		return nil, fferr.NewInternalError(fmt.Errorf("cannot lock an empty key"))
+	}
+
+	id := uuid.New().String()
+
+	// Get the lock
+	var existingLockID, existingResourceKey string
+	var existingTimestamp time.Time
+	selectQuery := l.getLockQuery()
+	err := l.db.QueryRow(context.Background(), selectQuery, key).Scan(&existingLockID, &existingResourceKey, &existingTimestamp)
+	if err != nil {
+		if strings.Contains(err.Error(), "no rows in result set") {
+			insertQuery := l.insertLockQuery()
```

Comment:
Race condition, we should check and insert in one transaction or in just a single query.